presto-hive/src/main/java/com/facebook/presto/hive/util/SerDeUtils.java

```diff
@@ -137,24 +137,27 @@ private static Slice serializeList(BlockBuilder builder, Object object, ListObje
             return null;
         }
 
+        List<Type> typeParameters = type.getTypeParameters();
+        checkArgument(typeParameters.size() == 1, "list must have exactly 1 type parameter");
+        Type elementType = typeParameters.get(0);
         ObjectInspector elementInspector = inspector.getListElementObjectInspector();
-        BlockBuilder currentBuilder = createBlockBuilder();
+        BlockBuilder currentBuilder = elementType.createBlockBuilder(new BlockBuilderStatus(), list.size());
 
         for (Object element : list) {
-            serializeObject(currentBuilder, element, elementInspector);
+            serializeObject(elementType, currentBuilder, element, elementInspector);
         }
 
-        SliceOutput out = new DynamicSliceOutput(1024);
-        currentBuilder.getEncoding().writeBlock(out, currentBuilder.build());
-
+        Block resultBlock = currentBuilder.build();
         if (builder != null) {
-            VARBINARY.writeSlice(builder, out.slice());
+            type.writeObject(builder, resultBlock);
+            return null;
```

Comment:
Only call-site that doesn't throw away the result use null as the argument for builder. In old code, it's free to return something. Now, it costs to return from `serializeRow`. And after your requested change above, it costs to return from `serializeList` and `serializeMap` as well.